lib/src/services/bedrock_service.dart

```diff
@@ -3,31 +3,67 @@ import 'package:dio/dio.dart';
 import 'package:flutter_secure_storage/flutter_secure_storage.dart';
 import 'package:jwt_decoder/jwt_decoder.dart';
 
+import 'dio_logger.dart';
+
+final Dio dio = Dio();
+
 class BedrockNetworkInterceptor extends Interceptor {
   final IAuthTokenStorage storage = AuthStorage(const FlutterSecureStorage());
 
+  static String TAG = "services_base_api";
+
   @override
   void onRequest(
     RequestOptions options,
     RequestInterceptorHandler handler,
   ) async {
-    String token = await storage.readAuthToken() ?? '';
+    String? token = await storage.readAuthToken(); //get token
     Map<String, dynamic> headers = {
       'Accept': 'application/json',
       'content-type': 'application/json',
     };
 
-    try {
+    if (token != null) {
       if (!JwtDecoder.isExpired(token)) {
         await storage.storeAuthToken(token);
         options.headers.addAll({...headers, 'Authorization': 'Bearer $token'});
+      } else {
+        dio.interceptors.requestLock
+            .lock(); // lock the request to wait for new token to be given
+        dio.interceptors.responseLock
+            .lock(); // lock the request to wait for new token to be given
```

Comment:
Just curious, since we're locking the request, do we also need to explicitly lock the response? I would imagine the response never coming through since the request was locked. I may be missing something but was just curious about this mechanism.

Reply:
for what I've read, it's good practice to look both, your question is very interesting and I asked myself the same when seeing others doing it, guess it's better to have more than less